### ISSUE TITLE
The aistore cluster's target node is not stable and is easily rejected by the cluster.

### ISSUE THREAD

**Issue by yingca1** [NVIDIA/aistore#134]
deploy method: `ais-k8s operator`

docker image:
- name: aistorage/ais-init
  newTag: latest
- name: aistorage/aisnode
  newTag: latest

error log:
```bash
aisnode args: -config=/etc/ais/ais.json -local_config=/etc/ais/ais_local.json -role=target -alsologtostderr=true -stderrthreshold=1  -allow_shared_no_disks=false
W 19:59:42.787163 config.go:1253 control and data share one intra-cluster network (aistore1-target-2.aistore1-target.ais.svc.cluster.local)
I 19:59:42.787625 config.go:1796 log.dir: "/var/log/ais"; l4.proto: tcp; pub port: 51081; verbosity: 3
I 19:59:42.787637 config.go:1798 config: "/etc/ais/.ais.conf"; stats_time: 10s; authentication: false; backends: [aws gcp]
I 19:59:42.787654 daemon.go:175 Version 3.17.66420ae, build time 2023-06-08T21:39:50+0000, debug false
I 19:59:42.787660 daemon.go:183 CPUs(4, runtime=4), containerized
I 19:59:42.787824 util.go:39 Verifying type of deployment (HOSTNAME: "aistore1-target-2", K8S_NODE_NAME: "")
I 19:59:42.807851 util.go:76 Successfully got node name "gke-ais1-ais1-5aa6ad88-p393", assuming Kubernetes deployment
I 19:59:42.807926 utils.go:149 Found only one IPv4: 10.72.7.9, MTU 1460
W 19:59:42.807935 utils.go:151 IPv4 10.72.7.9 MTU size is small: 1460
I 19:59:42.807947 htrun.go:346 PUBLIC (user) access: [{10.72.7.9 51081 http://10.72.7.9:51081 10.72.7.9:51081}]
I 19:59:42.807955 utils.go:112 Selecting one of the configured IPv4 addresses: [aistore1-target-2.aistore1-target.ais.svc.cluster.local]...
W 19:59:42.807959 utils.go:119 failed to parse IP for hostname "aistore1-target-2.aistore1-target.ais.svc.cluster.local"
W 19:59:42.808205 utils.go:132 Selected IPv4 10.72.7.9 from the configuration file
I 19:59:42.808221 htrun.go:359 INTRA-CONTROL access: [{aistore1-target-2.aistore1-target.ais.svc.cluster.local 51082 http://aistore1-target-2.aistore1-target.ais.svc.cluster.local:51082 aistore1-target-2.aistore1-target.ais.svc.cluster.local:51082}] (config: aistore1-target-2.aistore1-target.ais.svc.cluster.local)
I 19:59:42.808227 utils.go:112 Selecting one of the configured IPv4 addresses: [aistore1-target-2.aistore1-target.ais.svc.cluster.local]...
W 19:59:42.808232 utils.go:119 failed to parse IP for hostname "aistore1-target-2.aistore1-target.ais.svc.cluster.local"
W 19:59:42.808283 utils.go:132 Selected IPv4 10.72.7.9 from the configuration file
I 19:59:42.808292 htrun.go:372 INTRA-DATA access: [{aistore1-target-2.aistore1-target.ais.svc.cluster.local 51083 http://aistore1-target-2.aistore1-target.ais.svc.cluster.local:51083 aistore1-target-2.aistore1-target.ais.svc.cluster.local:51083}] (config: aistore1-target-2.aistore1-target.ais.svc.cluster.local)
I 19:59:42.808492 init.go:213 fHUWONyl.gmm[(used 4GiB, free 4GiB, buffcache 3GiB, actfree 7GiB), (min-free 2GiB, low-wm 4GiB), pressure 'low'] started
I 19:59:42.808599 init.go:213 fHUWONyl.smm[(used 4GiB, free 4GiB, buffcache 3GiB, actfree 7GiB), (min-free 2GiB, low-wm 4GiB), pressure 'low'] started
I 19:59:42.813384 dutils_linux.go:109 /dev/nvme0n2: map[nvme0n2:512]
I 19:59:42.813425 vinit.go:79 VMD v1(fHUWONyl, [/ais1])
I 19:59:42.813496 Using Prometheus
I 19:59:43.813746 fshc.go:60 Starting fshc
I 19:59:43.813770 collect.go:48 Intra-cluster networking: fasthttp client
I 19:59:43.813791 collect.go:49 Starting stream_collector
W 19:59:43.963008 bucketmeta.go:376 initializing new BMD v0
I 19:59:43.963036 etlmeta.go:221 initializing new EtlMD v0(0)
E 19:59:43.963996 htrun.go:429 FATAL ERROR: [bad cluster map]: t[fHUWONyl] is not present in the loaded Smap v109[klneDAlWs, p[oJDFAKuH], t=4, p=3]
FATAL ERROR: [bad cluster map]: t[fHUWONyl] is not present in the loaded Smap v109[klneDAlWs, p[oJDFAKuH], t=4, p=3]
cat: /var/log/ais/aisnode.INFO: No such file or directory
cat: /var/log/ais/aisnode.ERROR: No such file or directory
cat: /var/log/ais/aisnode.WARNING: No such file or directory
```

How do I appropriately handle this type of question?

**Comment by alex-aizman** [comment]
looks like this target somehow ended up with invalid cluster map - happens when you play with deployment and redeployment, so try to cleanup its metadata and restart

**Comment by alex-aizman** [comment]
separately. pub port 51081 - is it on purpose? suggesting to check everything related to networking configuration, including MTU - lookup "jumbo"

**Comment by yingca1** [comment]
> looks like this target somehow ended up with invalid cluster map - happens when you play with deployment and redeployment, so try to cleanup its metadata and restart

As it is a pod in Kubernetes, if the health check fails, it will keep restarting and repeating the error. Could you please clarify which files or operations "cleanup its metadata" refers to?

**Comment by yingca1** [comment]
> separately. pub port 51081 - is it on purpose? suggesting to check everything related to networking configuration, including MTU - lookup "jumbo"

`pub port 51081` refers to the code [here](https://github.com/NVIDIA/ais-k8s/blob/6ace0207028226796a55094a6dad5a6e40c447eb/docs/walkthrough.md#deploy-an-aistore-cluster). I'm not sure if it's correct or why it's configured as 51081. Thank you very much for your suggestion. I'll try configuring the MTU right away.

**Comment by saiprashanth173** [comment]
@yingca1 do you mind elaborating a bit on your setup and how we can reproduce this error?

I would be interested in knowing:
1. Where you are deploying your aistore cluster, e.g. minikube, baremetal K8s cluster or a public cloud K8s offering?
2. Do you see these logs in all your targets? 
3. Does it happen when you apply you create the AIStore cluster for the first time? Or you see it when you upgrade/patch existing cluster?

**Comment by yingca1** [comment]
@saiprashanth173 

1. Where you are deploying your aistore cluster, e.g. minikube, baremetal K8s cluster or a public cloud K8s offering?
> I launched and tested it on GKE using Terraform.

2. Do you see these logs in all your targets?
3. Does it happen when you apply you create the AIStore cluster for the first time? Or you see it when you upgrade/patch existing cluster?
> This error often occurs, for example, when initializing with 3 proxies and 3 targets, manually scaling the targets to 5 may very likely result in this error. Once, when I deleted all Kubernetes resources and redeployed, all targets reported similar errors.


Question:
Is there a way to manually add a target to the aistore cluster in a Kubernetes cluster? For example, if I increase the liveness time, can I manually fix the issue of not being able to join the cluster during that time?

**Comment by alex-aizman** [comment]
we are adding and removing targets all the time, it's part of the story since day one

* https://github.com/NVIDIA/aistore/blob/master/docs/lifecycle.md

**Comment by alex-aizman** [comment]
by way of background info: 

* when an ais target is deployed the first time, it records it's unique ID into each and every mountpath (grep docs for `mountpath`)
* the assumption and requirement is that this ID never disappears. In other words, the only legal way for it to disappear is decommission (ditto, docs)
* because, if it does, then the target will regenerate a new one and it'll collide with the cluster map (that is also very persistent, versioned, and replicated across)

**Comment by saiprashanth173** [comment]
I guess in your case `smap`, i.e. cluster map, from your previous deployment is causing your target to fail. 

One option would be to clean up metadata directory `/etc/ais` directory ([see](https://github.com/NVIDIA/ais-k8s/blob/6ace0207028226796a55094a6dad5a6e40c447eb/operator/config/samples/ais_v1beta1_aistore.yaml#L32)) on each node before you try to deploy a new aistore cluster and test it. Maybe it also makes sense to delete your old PV(C)s. Unfortunately, we don't have any helpers to do that on a GKE setup, so you might have to do it manually. [Here](https://github.com/NVIDIA/ais-k8s/blob/6ace0207028226796a55094a6dad5a6e40c447eb/playbooks/roles/ais_cleanup_all/tasks/main.yml) is the ansible role we use for a baremetal setup if you would like to take a look as a reference. 

Currently, we mostly use GKE in our CI, where we setup our GKE cluster using terraform, deploy aistore, run tests, which also include scale up (adding targets), scale down (removing targets) etc. After we run all our tests, we destroy the entire cluster. So, we never needed such utilities. 

For reference, some tests cases that run similar scenarios: 
https://github.com/NVIDIA/ais-k8s/blob/master/operator/tests/integration/cluster_test.go#L164
https://github.com/NVIDIA/ais-k8s/blob/master/operator/tests/integration/cluster_test.go#L240  

> Question:
Is there a way to manually add a target to the aistore cluster in a Kubernetes cluster? For example, if I increase the liveness time, can I manually fix the issue of not being able to join the cluster during that time?

Please let us know if cleaning metadata directory fixes your issue. 

Could you elaborate on how you plan to manually fixing the issue? In general, if your aistore cluster is being managed by ais operator, I would prefer using the scale up feature to add a new target instead trying to add it manually.

**Comment by yingca1** [comment]
Here's what I did and it seems like the problem hasn't occurred again: 

1. Upgraded GKE from version 1.25.8-gke.1000 (default) to 1.26.3-gke.1000 (I was originally planning to change the MTU, but it was mentioned that [GKE version 1.26.1 and later](https://cloud.google.com/kubernetes-engine/docs/concepts/network-overview#mtu ) can inherit from the primary interface of the Node). 

2. Added sysctls configuration in Terraform, following [this guide](https://github.com/NVIDIA/ais-k8s/blob/master/playbooks/vars/host_config_sysctl.yml). The commented out part is not supported by Terraform. 
```terraform
  ...
  node_config {
    linux_node_config {
      sysctls = {
        "net.core.somaxconn"                 = "100000"
        "net.ipv4.tcp_tw_reuse"              = "1"
        # "net.ipv4.ip_local_port_range"       = "2048 65535"
        # "net.ipv4.tcp_max_tw_buckets"        = "1440000"
        # "net.ipv4.ip_forward"                = "1"
        "net.core.rmem_max"                  = "268435456"
        "net.core.wmem_max"                  = "268435456"
        # "net.core.rmem_default"              = "25165824"
        "net.core.optmem_max"                = "25165824"
        "net.core.netdev_max_backlog"        = "250000"
        "net.ipv4.tcp_wmem"                  = "4096    12582912  268435456"
        "net.ipv4.tcp_rmem"                  = "4096    12582912 268435456"
        # "net.ipv4.tcp_adv_win_scale"         = "1"
        # "net.ipv4.tcp_mtu_probing"           = "2"
        # "net.ipv4.tcp_slow_start_after_idle" = "0"
        # "net.ipv4.tcp_low_latency"           = "1"
        # "net.ipv4.tcp_timestamps"            = "0"
        # "vm.vfs_cache_pressure"              = "50"
        # "net.ipv4.tcp_max_syn_backlog"       = "100000"
        # "net.ipv4.tcp_rfc1337"               = "1"
        # "vm.swappiness"                      = "10"
        # "vm.min_free_kbytes"                 = "262144"
      }
    }
    ...
  }
```


3. Destroyed the old Kubernetes cluster and reconfigured it.

**Comment by yingca1** [comment]
> Could you elaborate on how you plan to manually fix the issue? In general, if your aistore cluster is being managed by ais operator, I would prefer using the scale up feature to add a new target instead trying to add it manually.

@saiprashanth173 To be honest, after creating a complete aistore cluster using the operator, I dumped all the resources of the ais-operator-system in Kubernetes into a YAML file. This allowed me to quickly debug by directly modifying the YAML resource file

**Comment by yingca1** [comment]
This issue has come up again, but I think I've found the root cause. 

TODO: Please wait for me to update here later

**Comment by saiprashanth173** [comment]
> @saiprashanth173 To be honest, after creating a complete aistore cluster using the operator, I dumped all the resources of the ais-operator-system in Kubernetes into a YAML file. This allowed me to quickly debug by directly modifying the YAML resource file

I am a bit curious how fixing the YAML resources fixed this issue. To me from the error log you shared, this seems to be storage target issue, and has less to do with Kubernetes resources itself.  

> E 19:59:43.963996 htrun.go:429 FATAL ERROR: [bad cluster map]: t[fHUWONyl] is not present in the loaded Smap v109[klneDAlWs, p[oJDFAKuH], t=4, p=3]
FATAL ERROR: [bad cluster map]: t[fHUWONyl] is not present in the loaded Smap v109[klneDAlWs, p[oJDFAKuH], t=4, p=3]

It would be helpful if you can also share what you had to change in YAMLs to be to able to debug and fix this issue. We can use that to decide if we need to fix/extend the K8s operator. 

> This issue has come up again, but I think I've found the root cause.
>
> TODO: Please wait for me to update here later

That is amazing! Will wait for you update :slightly_smiling_face:. Thanks!

**Comment by alex-aizman** [comment]
maybe a little bit more info on the original error, and I'm not sure you need it at this point: 

```console
I 19:59:42.787654 daemon.go:175 Version 3.17.66420ae, build time 2023-06-08T21:39:50+0000, debug false
```
First, there's the version. It's a not yet released v3.18 which's in progress, changing fast, and a bit risky to use. We typically release aistore and ais operator more or less at the same time, as a package. 

Secondly, there's the question of persistence:

```console
FATAL ERROR: [bad cluster map]: t[fHUWONyl] is not present in the loaded Smap v109[klneDAlWs, p[oJDFAKuH], t=4, p=3]
```

On the one hand, cluster map (aka `Smap`) is stored in the host under `/etc/ais`. On the other hand, target `ID` is stored at the root of each `mountpath` on data drives that Kubernetes calls "persistent volumes". One would hope, they are persistent enough and have the right affinity, etc. Something to maybe check.
